programs/quarter_board/main.c

```diff
@@ -0,0 +1,192 @@
+#include "ch.h"
+#include "hal.h"
+
+#include <icu_lld.h>
+#include "chprintf.h"
+#include <inttypes.h>
+
+
+#include "communication.h"
+
+#define BIG_BOOLSHIT 3000
+#define SMALL_BOOLSHIT 400
+#define NOT_SEEING_VALUE 3000
+#define QUANTITY_OF_SENSORS 2
+
+
+int32_t send_values[QUANTITY_OF_SENSORS];
+int32_t sensors_values[QUANTITY_OF_SENSORS];
+bool values_available[QUANTITY_OF_SENSORS];
+
+thread_t * commanderp;
+thread_t * ir_measurementp;
+
+/*
+ * ICU callbacks
+ */
+
+static void read_pulse(ICUDriver *icup) {
+    uint8_t i;
+    int32_t value;
+
+    if(icup == &ICUD1) {
+        i = 0;
+    } else {
+        i = 1;
+    }
+    
+    value = icuGetWidthX(icup);
+
+    if(value < BIG_BOOLSHIT && value > SMALL_BOOLSHIT) {    
+        sensors_values[i] = value;
+        values_available[i] = true;
+    } else {
+        values_available[i] = false;
+    }
+}
+
+THD_WORKING_AREA(waThread1, 64);
+THD_FUNCTION(Thread1, arg) {
+    (void)arg;
+
+    uint8_t i;
+
+    while (1) {
+        if(i_am_middle_one) {
+            values_available[0] = false;
+        }
+
+        for(i = 0; i < QUANTITY_OF_SENSORS; i++) {
+            if(values_available[i]) {
+                send_values[i] = sensors_values[i];
+                values_available[i] = false;
+            } else {
+                send_values[i] = NOT_SEEING_VALUE;
+            }
+        }
+        chThdSleepMilliseconds(100);
+    }
+    chRegSetThreadName("ir_measurement");
+    chThdSleepMilliseconds(100);
+}
+void execute_master_command(uint16_t command_id, uint8_t data_lenght, uint8_t *buff) {
+    switch(command_id) {
+        case 0xAA:  if(buff[0] == 1) led(GREEN, 1);
+                    else led(GREEN, 0);
+
+                    if(buff[1] == 1) led(RED, 1);
+                    else led(RED, 0);
+
+                    break;
+    }
+
+}
+
+uint8_t *send_data_command(uint16_t command_id, uint8_t data_lenght, uint8_t *buff) {
+    int32_t values = 0;
+    uint8_t i;
+    switch(command_id) {
+        case 0xFF: values = send_values[0] | (send_values[1] << 16);
+                   led(GREEN, 1);
+                   led(RED, 0);
+                   for(i = 0; i < data_lenght; i++) {
+                       buff[i] = ((values >> (4 * i)) & 0xf);
+                   }
+                   break; 
+    }
+    return buff;
+}
+
+THD_WORKING_AREA(waCommanderThread, 128);
+THD_FUNCTION(CommanderThread, arg) {
+    (void)arg;
+
+    thread_t * serialp;
+    MESSAGE *messagep;
+    uint16_t cmd, command_id;
+    uint8_t data_buff[MAX_DATA_CHAR_LENGHT];
+    uint8_t data_lenght, i, rw_bit;;
+    while(1) {
+        serialp = chMsgWait();
+        messagep =(MESSAGE *) chMsgGet(serialp);
+        
+        cmd = messagep -> operation;
+
+        command_id = (cmd & 0b11111111111);

```

Comment:
I think in this case it's better to use bin rather than hex too because command_id is one part of 16bit command which contain 4bit data lenght, 1 read-write operation bit, and 11bit command id. I think this is better for identify "bit size" of command_id than hex value.